File: sourcegen-generator-java/src/test/java/io/micronaut/sourcegen/javapoet/write/GeneratedProgramTest.java
/*
 * Copyright 2017-2026 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.micronaut.sourcegen.javapoet.write;

import io.micronaut.sourcegen.model.ClassDef;
import io.micronaut.sourcegen.model.ClassTypeDef;
import io.micronaut.sourcegen.model.EnumDef;
import io.micronaut.sourcegen.model.ExpressionDef;
import io.micronaut.sourcegen.model.FieldDef;
import io.micronaut.sourcegen.model.MethodDef;
import io.micronaut.sourcegen.model.PropertyDef;
import io.micronaut.sourcegen.model.RecordDef;
import io.micronaut.sourcegen.model.StatementDef;
import io.micronaut.sourcegen.model.TypeDef;
import org.junit.jupiter.api.Test;

import javax.lang.model.element.Modifier;
import java.lang.reflect.Array;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Function;
import java.util.function.Supplier;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import static io.micronaut.sourcegen.javapoet.write.JavaCompileAssertions.compileMatchingSnapshots;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertThrows;

/** Executes DSL-generated programs across generic, control-flow, and receiver-capture boundaries. */
class GeneratedProgramTest {

    private static final String SNAPSHOTS = "/generated-programs/java";

    @ParameterizedTest(name = "a generic array of rank {0} keeps its values through the call")
    @ValueSource(ints = {1, 2, 3})
    void genericArrayCallsPreserveValuesAcrossRanks(int rank) throws Exception {
        var t = TypeDef.variable("T", TypeDef.of(Integer.class));
        var identity = MethodDef.builder("identity").addModifiers(Modifier.PUBLIC).addTypeVariable(t)
            .addParameter("value", t.array(rank)).returns(t.array(rank))
            .build((self, p) -> p.getFirst().returning());
        var def = ClassDef.builder("test.Rank" + rank).addModifiers(Modifier.PUBLIC).addMethod(identity)
            .addMethod(MethodDef.builder("call").addModifiers(Modifier.PUBLIC)
                .addParameter("value", TypeDef.of(Number.class).array(rank)).returns(Object.class)
                .build((self, p) -> self.invoke(identity, p.getFirst()).returning())).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, def)) {
            var cls = loader.loadClass(def.getName());
            Object value = Array.newInstance(Integer.class, new int[rank]);
            Class<?> parameter = Array.newInstance(Number.class, new int[rank]).getClass();
            assertSame(value, cls.getMethod("call", parameter).invoke(cls.getConstructor().newInstance(), value));
        }
    }

    @ParameterizedTest(name = "a chain of {0} bounds keeps the runtime identity of the value")
    @ValueSource(ints = {1, 2, 8, 9})
    void chainedBoundsKeepTheirRuntimeIdentity(int length) throws Exception {
        var variables = new ArrayList<TypeDef.TypeVariable>();
        TypeDef bound = TypeDef.of(CharSequence.class);
        for (int i = length - 1; i >= 0; i--) {
            var variable = TypeDef.variable("T" + i, bound);
            variables.addFirst(variable);
            bound = variable;
        }
        var method = MethodDef.builder("identity").addModifiers(Modifier.PUBLIC);
        variables.forEach(method::addTypeVariable);
        var identity = method.addParameter("value", variables.getFirst()).returns(variables.getFirst())
            .build((self, p) -> p.getFirst().returning());
        var def = ClassDef.builder("test.Chain" + length).addModifiers(Modifier.PUBLIC).addMethod(identity)
            .addMethod(MethodDef.builder("call").addModifiers(Modifier.PUBLIC).addParameter("value", Object.class).returns(Object.class)
                .build((self, p) -> self.invoke(identity, p.getFirst()).returning())).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, def)) {
            var cls = loader.loadClass(def.getName());
            assertEquals("text", cls.getMethod("call", Object.class).invoke(cls.getConstructor().newInstance(), "text"));
        }
    }

    @ParameterizedTest(name = "a reference through a {0} captures its receiver before it changes")
    @ValueSource(strings = {"field", "local", "result"})
    void referencesCaptureTheirReceiverBeforeItChanges(String kind) throws Exception {
        var label = FieldDef.builder("label", String.class).addModifiers(Modifier.PRIVATE).build();
        var apply = MethodDef.builder("apply").addModifiers(Modifier.PUBLIC).overrides().addParameter("value", Object.class).returns(Object.class)
            .build((self, p) -> self.field(label).returning());
        var target = ClassDef.builder("test.CapturedTarget").addModifiers(Modifier.PUBLIC).addField(label)
            .addAllFieldsConstructor(Modifier.PUBLIC).addSuperinterface(TypeDef.parameterized(Function.class, String.class, String.class)).addMethod(apply).build();
        var receiver = FieldDef.builder("receiver", target.asTypeDef()).addModifiers(Modifier.PRIVATE)
            .initializer(target.asTypeDef().instantiate(ExpressionDef.constant("first"))).build();
        var count = FieldDef.builder("count", int.class).addModifiers(Modifier.PRIVATE).initializer(ExpressionDef.constant(0)).build();
        var next = MethodDef.builder("next").addModifiers(Modifier.PUBLIC).returns(target.asTypeDef())
            .build((self, p) -> StatementDef.multi(self.field(count).put(self.field(count).math(ExpressionDef.MathBinaryOperation.OpType.ADDITION, ExpressionDef.constant(1))), self.field(receiver).returning()));
        var function = TypeDef.parameterized(Function.class, Object.class, Object.class);
        var caller = ClassDef.builder("test.Capture" + kind).addModifiers(Modifier.PUBLIC).addField(receiver).addField(count).addMethod(next)
            .addMethod(MethodDef.builder("reads").addModifiers(Modifier.PUBLIC).returns(int.class).build((self, p) -> self.field(count).returning()))
            .addMethod(MethodDef.builder("replace").addModifiers(Modifier.PUBLIC).returns(void.class)
                .build((self, p) -> self.field(receiver).put(target.asTypeDef().instantiate(ExpressionDef.constant("second")))))
            .addMethod(MethodDef.builder("capture").addModifiers(Modifier.PUBLIC).returns(function).build((self, p) -> switch (kind) {
                case "field" -> function.methodReference(self.field(receiver), apply).returning();
                case "local" -> self.field(receiver).newLocal("target", local -> function.methodReference(local, apply).returning());
                default -> function.methodReference(self.invoke(next), apply).returning();
            })).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, target, caller)) {
            var cls = loader.loadClass(caller.getName());
            var instance = cls.getConstructor().newInstance();
            @SuppressWarnings("unchecked") var captured = (Function<Object, Object>) cls.getMethod("capture").invoke(instance);
            cls.getMethod("replace").invoke(instance);
            assertEquals("first", captured.apply("one"));
            assertEquals("first", captured.apply("two"));
            assertEquals(kind.equals("result") ? 1 : 0, cls.getMethod("reads").invoke(instance));
        }
    }

    @Test
    void genericRecordOverrideReturnsTheComponent() throws Exception {
        var t = TypeDef.variable("T", TypeDef.of(CharSequence.class));
        var record = RecordDef.builder("test.SuppliedRecord").addModifiers(Modifier.PUBLIC).addTypeVariable(t)
            .addProperty(PropertyDef.builder("value").ofType(t).build())
            .addSuperinterface(TypeDef.parameterized(ClassTypeDef.of(Supplier.class), t))
            .addMethod(MethodDef.builder("get").addModifiers(Modifier.PUBLIC).overrides().returns(Object.class)
                .build((self, p) -> self.field("value", t).returning())).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, record)) {
            var cls = loader.loadClass(record.getName());
            assertEquals("text", ((Supplier<?>) cls.getConstructor(CharSequence.class).newInstance("text")).get());
        }
    }

    @Test
    void enumOverrideIsCallableThroughItsGenericInterface() throws Exception {
        var def = EnumDef.builder("test.SuppliedEnum").addModifiers(Modifier.PUBLIC).addEnumConstant("ONE")
            .addSuperinterface(TypeDef.parameterized(Supplier.class, String.class))
            .addMethod(MethodDef.builder("get").addModifiers(Modifier.PUBLIC).overrides().returns(Object.class)
                .build((self, p) -> ExpressionDef.constant("one").returning())).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, def)) {
            assertEquals("one", ((Supplier<?>) loader.loadClass(def.getName()).getField("ONE").get(null)).get());
        }
    }

    @Test
    void typedConstructorConvertsAnErasedArgument() throws Exception {
        var constructor = StringBuilder.class.getConstructor(CharSequence.class);
        var def = ClassDef.builder("test.TypedConstructor").addModifiers(Modifier.PUBLIC)
            .addMethod(MethodDef.builder("create").addModifiers(Modifier.PUBLIC).addParameter("value", Object.class).returns(StringBuilder.class)
                .build((self, p) -> ClassTypeDef.of(StringBuilder.class).instantiate(constructor, p.getFirst()).returning())).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, def)) {
            var cls = loader.loadClass(def.getName());
            assertEquals("text", cls.getMethod("create", Object.class).invoke(cls.getConstructor().newInstance(), "text").toString());
        }
    }

    @Test
    void finallyRunsOnceOnBothReturnAndThrow() throws Exception {
        var increment = AtomicInteger.class.getMethod("incrementAndGet");
        var def = ClassDef.builder("test.FinallyPaths").addModifiers(Modifier.PUBLIC)
            .addMethod(MethodDef.builder("call").addModifiers(Modifier.PUBLIC).addParameter("counter", AtomicInteger.class)
                .addParameter("fail", boolean.class).returns(int.class)
                .build((self, p) -> StatementDef.doTry(StatementDef.multi(
                    p.get(1).isTrue().doIf(ClassTypeDef.of(IllegalStateException.class).instantiate().doThrow()), ExpressionDef.constant(7).returning()
                )).doFinally(p.getFirst().invoke(increment)))).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, def)) {
            var cls = loader.loadClass(def.getName());
            var instance = cls.getConstructor().newInstance();
            var method = cls.getMethod("call", AtomicInteger.class, boolean.class);
            var counter = new AtomicInteger();
            assertEquals(7, method.invoke(instance, counter, false));
            var failure = assertThrows(InvocationTargetException.class, () -> method.invoke(instance, counter, true));
            assertInstanceOf(IllegalStateException.class, failure.getCause());
            assertEquals(2, counter.get());
        }
    }

    @ParameterizedTest(name = "a static final field assigned through {0} is initialized once")
    @ValueSource(strings = {"ifTrue", "ifFalse", "switchAll", "switchPartial", "loop", "monitor"})
    void staticInitializationHandlesEveryAssignmentPath(String kind) throws Exception {
        var owner = ClassTypeDef.of("test.Initialize" + kind);
        var field = FieldDef.builder("value", int.class).addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL).build();
        var assign = owner.getStaticField(field).put(ExpressionDef.constant(7));
        StatementDef body = switch (kind) {
            case "ifTrue" -> ExpressionDef.trueValue().isTrue().doIf(assign);
            case "ifFalse" -> ExpressionDef.falseValue().isTrue().doIf(assign);
            case "switchAll" -> ExpressionDef.constant(1).asStatementSwitch(TypeDef.Primitive.INT,
                Map.of(ExpressionDef.constant(1), assign), owner.getStaticField(field).put(ExpressionDef.constant(9)));
            case "switchPartial" -> ExpressionDef.constant(0).asStatementSwitch(TypeDef.Primitive.INT,
                Map.of(ExpressionDef.constant(1), assign), StatementDef.multi());
            case "loop" -> ExpressionDef.constant(0).newLocal("index", local -> local.compare(ExpressionDef.ComparisonOperation.OpType.LESS_THAN, ExpressionDef.constant(2))
                .whileLoop(StatementDef.multi(owner.getStaticField(field).put(local), local.assign(local.math(ExpressionDef.MathBinaryOperation.OpType.ADDITION, ExpressionDef.constant(1))))));
            default -> new StatementDef.Synchronized(ExpressionDef.constant("lock"), assign);
        };
        var def = ClassDef.builder(owner.getName()).addModifiers(Modifier.PUBLIC).addField(field).addStaticInitializer(body).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, def)) {
            int expected = switch (kind) { case "ifFalse", "switchPartial" -> 0; case "loop" -> 1; default -> 7; };
            assertEquals(expected, loader.loadClass(def.getName()).getField("value").get(null));
        }
    }

    @ParameterizedTest(name = "a local declared in a {0} cannot capture the static field assignment")
    @ValueSource(strings = {"if", "switch", "loop", "monitor"})
    void nestedLocalNamesCannotCaptureAStaticFieldAssignment(String kind) throws Exception {
        var owner = ClassTypeDef.of("test.Shadow" + kind);
        var field = FieldDef.builder("value", int.class).addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL).build();
        var local = ExpressionDef.constant(1).newLocal("value", ignored -> StatementDef.multi());
        var parse = Boolean.class.getMethod("parseBoolean", String.class);
        StatementDef nested = switch (kind) {
            case "if" -> ExpressionDef.trueValue().isTrue().doIf(local);
            case "switch" -> ExpressionDef.constant(0).asStatementSwitch(TypeDef.Primitive.INT, Map.of(ExpressionDef.constant(0), local), StatementDef.multi());
            case "loop" -> ClassTypeDef.of(Boolean.class).invokeStatic(parse, ExpressionDef.constant("false")).isTrue().whileLoop(local);
            default -> new StatementDef.Synchronized(ExpressionDef.constant("lock"), local);
        };
        var def = ClassDef.builder(owner.getName()).addModifiers(Modifier.PUBLIC).addField(field)
            .addStaticInitializer(StatementDef.multi(nested, owner.getStaticField(field).put(ExpressionDef.constant(9)))).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, def)) {
            assertEquals(9, loader.loadClass(def.getName()).getField("value").get(null));
        }
    }

    @Test
    void arrayReceiverArgumentsResolveVariablesOutsideTheCallerScope() throws Exception {
        var t = TypeDef.variable("T");
        var stored = FieldDef.builder("stored", Object.class).addModifiers(Modifier.PRIVATE).build();
        var get = MethodDef.builder("get").addModifiers(Modifier.PUBLIC).overrides().returns(Object.class)
            .build((self, p) -> self.field(stored).returning());
        var target = ClassDef.builder("test.ArrayReceiver").addModifiers(Modifier.PUBLIC).addTypeVariable(t).addField(stored)
            .addAllFieldsConstructor(Modifier.PUBLIC)
            .addSuperinterface(TypeDef.parameterized(ClassTypeDef.of(Supplier.class), t)).addMethod(get).build();
        var outOfScope = TypeDef.variable("U", TypeDef.of(CharSequence.class)).array();
        var caller = ClassDef.builder("test.ArrayReceiverCaller").addModifiers(Modifier.PUBLIC)
            .addMethod(MethodDef.builder("call").addModifiers(Modifier.PUBLIC).addParameter("target", TypeDef.parameterized(target.asTypeDef(), outOfScope))
                .returns(Object.class).build((self, p) -> p.getFirst().invoke(get).returning())).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, target, caller)) {
            var targetClass = loader.loadClass(target.getName());
            var cls = loader.loadClass(caller.getName());
            var value = new String[]{"text"};
            assertSame(value, cls.getMethod("call", targetClass)
                .invoke(cls.getConstructor().newInstance(), targetClass.getConstructor(Object.class).newInstance((Object) value)));
        }
    }

    @Test
    void methodReferenceConvertsParameterizedArrayResults() throws Exception {
        var values = FieldDef.builder("values", Object.class).addModifiers(Modifier.PRIVATE).build();
        var get = MethodDef.builder("get").addModifiers(Modifier.PUBLIC).overrides().returns(Object.class)
            .build((self, p) -> self.field(values).returning());
        var target = ClassDef.builder("test.ArraySupplier").addModifiers(Modifier.PUBLIC).addField(values).addAllFieldsConstructor(Modifier.PUBLIC)
            .addSuperinterface(TypeDef.parameterized(ClassTypeDef.of(Supplier.class), TypeDef.parameterized(List.class, String.class).array())).addMethod(get).build();
        var result = TypeDef.parameterized(ClassTypeDef.of(Supplier.class), TypeDef.parameterized(List.class, Object.class).array());
        var caller = ClassDef.builder("test.ArraySupplierCaller").addModifiers(Modifier.PUBLIC)
            .addMethod(MethodDef.builder("reference").addModifiers(Modifier.PUBLIC).addParameter("target", target.asTypeDef()).returns(result)
                .build((self, p) -> result.methodReference(p.getFirst(), get).returning())).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, target, caller)) {
            var targetClass = loader.loadClass(target.getName());
            var cls = loader.loadClass(caller.getName());
            var value = new List<?>[]{List.of("text")};
            var instance = targetClass.getConstructor(Object.class).newInstance((Object) value);
            var reference = (Supplier<?>) cls.getMethod("reference", targetClass).invoke(cls.getConstructor().newInstance(), instance);
            assertSame(value, reference.get());
        }
    }

    @Test
    void exhaustiveConditionalDropsTheUnreachableFallback() throws Exception {
        var def = ClassDef.builder("test.ExhaustiveBranches").addModifiers(Modifier.PUBLIC)
            .addMethod(MethodDef.builder("call").addModifiers(Modifier.PUBLIC).addParameter("flag", boolean.class).returns(String.class)
                .build((self, p) -> StatementDef.multi(p.getFirst().isTrue().doIfElse(
                    ExpressionDef.constant("yes").returning(), ExpressionDef.constant("no").returning()), ExpressionDef.constant("unreachable").returning()))).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, def)) {
            var cls = loader.loadClass(def.getName());
            var instance = cls.getConstructor().newInstance();
            assertEquals("yes", cls.getMethod("call", boolean.class).invoke(instance, true));
            assertEquals("no", cls.getMethod("call", boolean.class).invoke(instance, false));
        }
    }

    @Test
    void enumCatchVariableDoesNotShadowItsExceptionField() throws Exception {
        var field = FieldDef.builder("e0", Throwable.class).addModifiers(Modifier.PRIVATE).build();
        var def = EnumDef.builder("test.CatchingEnum").addModifiers(Modifier.PUBLIC).addEnumConstant("ONE").addField(field)
            .addMethod(MethodDef.builder("call").addModifiers(Modifier.PUBLIC).returns(Throwable.class).build((self, p) -> StatementDef.multi(
                StatementDef.doTry(ClassTypeDef.of(IllegalStateException.class).instantiate(ExpressionDef.constant("failure")).doThrow())
                    .doCatch(Throwable.class, exception -> self.field(field).put(exception)), self.field(field).returning()))).build();
        try (var loader = compileMatchingSnapshots(SNAPSHOTS, def)) {
            var cls = loader.loadClass(def.getName());
            var value = (Throwable) cls.getMethod("call").invoke(cls.getField("ONE").get(null));
            assertInstanceOf(IllegalStateException.class, value);
            assertEquals("failure", value.getMessage());
        }
    }
}

File: sourcegen-generator-kotlin/src/test/kotlin/io/micronaut/sourcegen/GeneratedProgramTest.kt
/*
 * Copyright 2017-2026 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.micronaut.sourcegen

import io.micronaut.sourcegen.KotlinCompileAssertions.compileMatchingSnapshots
import io.micronaut.sourcegen.model.ClassDef
import io.micronaut.sourcegen.model.ClassTypeDef
import io.micronaut.sourcegen.model.EnumDef
import io.micronaut.sourcegen.model.ExpressionDef
import io.micronaut.sourcegen.model.FieldDef
import io.micronaut.sourcegen.model.MethodDef
import io.micronaut.sourcegen.model.StatementDef
import io.micronaut.sourcegen.model.TypeDef
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertInstanceOf
import org.junit.jupiter.api.Assertions.assertSame
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import java.lang.reflect.Array as ReflectArray
import java.lang.reflect.InvocationTargetException
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Function
import java.util.function.Supplier
import javax.lang.model.element.Modifier
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource

/** Executes DSL-generated programs, including evaluation order and generic conversions. */
class GeneratedProgramTest {

    private val snapshots = "/generated-programs/kotlin"

    @ParameterizedTest(name = "a generic array of rank {0} keeps its values through the call")
    @ValueSource(ints = [1, 2, 3])
    fun genericArrayCallsPreserveValuesAcrossRanks(rank: Int) {
        val t = TypeDef.variable("T", TypeDef.of(CharSequence::class.java))
        val identity = MethodDef.builder("array$rank").addModifiers(Modifier.PUBLIC, Modifier.STATIC).addTypeVariable(t)
            .addParameter("value", t.array(rank)).returns(t.array(rank)).build()
        val def = ClassDef.builder("test.Rank$rank")
            .addMethod(MethodDef.builder("call").addParameter("value", TypeDef.OBJECT.array(rank)).returns(TypeDef.OBJECT)
                .build { _, p -> ClassTypeDef.of(Calls::class.java).invokeStatic(identity, p[0]).returning() }).build()
        compileMatchingSnapshots(snapshots, def).use { loader ->
            val cls = loader.loadClass(def.name)
            val value = ReflectArray.newInstance(CharSequence::class.java, *IntArray(rank))
            val parameter = ReflectArray.newInstance(Any::class.java, *IntArray(rank)).javaClass
            assertSame(value, cls.getMethod("call", parameter).invoke(cls.getConstructor().newInstance(), value))
        }
    }

    @ParameterizedTest(name = "a compatible {0} argument keeps its value")
    @ValueSource(strings = ["number", "text", "list"])
    fun compatibleArgumentsKeepTheirValues(kind: String) {
        val bound = when (kind) {
            "number" -> TypeDef.of(Number::class.java)
            "text" -> TypeDef.of(CharSequence::class.java)
            else -> TypeDef.parameterized(List::class.java, String::class.java)
        }
        val input = when (kind) {
            "number" -> TypeDef.Primitive.INT
            "text" -> TypeDef.STRING
            else -> bound
        }
        val t = TypeDef.variable("T", bound)
        val identity = MethodDef.builder(kind).addModifiers(Modifier.PUBLIC, Modifier.STATIC).addTypeVariable(t)
            .addParameter("value", t).returns(t).build()
        val def = ClassDef.builder("test.Compatible$kind")
            .addMethod(MethodDef.builder("call").addParameter("value", input).returns(TypeDef.OBJECT)
                .build { _, p -> ClassTypeDef.of(Calls::class.java).invokeStatic(identity, p[0]).returning() }).build()
        compileMatchingSnapshots(snapshots, def).use { loader ->
            val cls = loader.loadClass(def.name)
            val parameter = when (kind) { "number" -> Int::class.javaPrimitiveType!!; "text" -> String::class.java; else -> List::class.java }
            val value: Any = when (kind) { "number" -> 7; "text" -> "text"; else -> listOf("text") }
            assertEquals(value, cls.getMethod("call", parameter).invoke(cls.getConstructor().newInstance(), value))
        }
    }

    @ParameterizedTest(name = "a reference through a {0} captures its receiver before it changes")
    @ValueSource(strings = ["field", "local", "result"])
    fun referencesCaptureTheirReceiverBeforeItChanges(kind: String) {
        val label = FieldDef.builder("label", String::class.java).addModifiers(Modifier.PRIVATE).build()
        val apply = MethodDef.builder("apply").addModifiers(Modifier.PUBLIC).overrides().addParameter("value", TypeDef.OBJECT).returns(TypeDef.OBJECT)
            .build { self, _ -> self.field(label).returning() }
        val target = ClassDef.builder("test.CapturedTarget").addField(label).addAllFieldsConstructor(Modifier.PUBLIC)
            .addSuperinterface(TypeDef.parameterized(Function::class.java, String::class.java, String::class.java)).addMethod(apply).build()
        val receiver = FieldDef.builder("receiver", target.asTypeDef()).addModifiers(Modifier.PRIVATE)
            .initializer(target.asTypeDef().instantiate(ExpressionDef.constant("first"))).build()
        val count = FieldDef.builder("count", TypeDef.Primitive.INT).addModifiers(Modifier.PRIVATE).initializer(ExpressionDef.constant(0)).build()
        val next = MethodDef.builder("next").returns(target.asTypeDef()).build { self, _ -> StatementDef.multi(
            self.field(count).put(self.field(count).math(ExpressionDef.MathBinaryOperation.OpType.ADDITION, ExpressionDef.constant(1))),
            self.field(receiver).returning()) }
        val function = TypeDef.parameterized(Function::class.java, Any::class.java, Any::class.java)
        val caller = ClassDef.builder("test.Capture$kind").addField(receiver).addField(count).addMethod(next)
            .addMethod(MethodDef.builder("reads").returns(TypeDef.Primitive.INT).build { self, _ -> self.field(count).returning() })
            .addMethod(MethodDef.builder("replace").returns(TypeDef.VOID)
                .build { self, _ -> self.field(receiver).put(target.asTypeDef().instantiate(ExpressionDef.constant("second"))) })
            .addMethod(MethodDef.builder("capture").returns(function).build { self, _ -> when (kind) {
                "field" -> function.methodReference(self.field(receiver), apply).returning()
                "local" -> self.field(receiver).newLocal("target") { local -> function.methodReference(local, apply).returning() }
                else -> function.methodReference(self.invoke(next), apply).returning()
            } }).build()
        compileMatchingSnapshots(snapshots, target, caller).use { loader ->
            val cls = loader.loadClass(caller.name)
            val instance = cls.getConstructor().newInstance()
            @Suppress("UNCHECKED_CAST")
            val captured = cls.getMethod("capture").invoke(instance) as Function<Any, Any>
            cls.getMethod("replace").invoke(instance)
            assertEquals("first", captured.apply("one"))
            assertEquals("first", captured.apply("two"))
            assertEquals(if (kind == "result") 1 else 0, cls.getMethod("reads").invoke(instance))
        }
    }

    @Test
    fun enumOverrideIsCallableThroughItsGenericInterface() {
        val def = EnumDef.builder("test.SuppliedEnum").addEnumConstant("ONE")
            .addSuperinterface(TypeDef.parameterized(Supplier::class.java, String::class.java))
            .addMethod(MethodDef.builder("get").addModifiers(Modifier.PUBLIC).overrides().returns(TypeDef.OBJECT)
                .build { _, _ -> ExpressionDef.constant("one").returning() }).build()
        compileMatchingSnapshots(snapshots, def).use { loader ->
            assertEquals("one", (loader.loadClass(def.name).getField("ONE").get(null) as Supplier<*>).get())
        }
    }

    @Test
    fun typedConstructorConvertsAnErasedArgument() {
        val constructor = StringBuilder::class.java.getConstructor(CharSequence::class.java)
        val def = ClassDef.builder("test.TypedConstructor")
            .addMethod(MethodDef.builder("create").addParameter("value", TypeDef.OBJECT).returns(StringBuilder::class.java)
                .build { _, p -> ClassTypeDef.of(StringBuilder::class.java).instantiate(constructor, p[0]).returning() }).build()
        compileMatchingSnapshots(snapshots, def).use { loader ->
            val cls = loader.loadClass(def.name)
            assertEquals("text", cls.getMethod("create", Any::class.java).invoke(cls.getConstructor().newInstance(), "text").toString())
        }
    }

    @Test
    fun finallyRunsOnceOnBothReturnAndThrow() {
        val increment = AtomicInteger::class.java.getMethod("incrementAndGet")
        val def = ClassDef.builder("test.FinallyPaths")
            .addMethod(MethodDef.builder("call").addParameter("counter", AtomicInteger::class.java)
                .addParameter("fail", TypeDef.Primitive.BOOLEAN).returns(TypeDef.Primitive.INT)
                .build { _, p -> StatementDef.doTry(StatementDef.multi(
                    p[1].isTrue().doIf(ClassTypeDef.of(IllegalStateException::class.java).instantiate().doThrow()), ExpressionDef.constant(7).returning()
                )).doFinally(p[0].invoke(increment)) }).build()
        compileMatchingSnapshots(snapshots, def).use { loader ->
            val cls = loader.loadClass(def.name)
            val instance = cls.getConstructor().newInstance()
            val method = cls.getMethod("call", AtomicInteger::class.java, Boolean::class.javaPrimitiveType)
            val counter = AtomicInteger()
            assertEquals(7, method.invoke(instance, counter, false))
            val failure = assertThrows(InvocationTargetException::class.java) { method.invoke(instance, counter, true) }
            assertInstanceOf(IllegalStateException::class.java, failure.cause)
            assertEquals(2, counter.get())
        }
    }

    @Test
    fun methodReferenceConvertsParameterizedArrayResults() {
        val values = FieldDef.builder("values", TypeDef.OBJECT).addModifiers(Modifier.PRIVATE).build()
        val get = MethodDef.builder("get").addModifiers(Modifier.PUBLIC).overrides().returns(TypeDef.OBJECT)
            .build { self, _ -> self.field(values).returning() }
        val target = ClassDef.builder("test.ArraySupplier").addField(values).addAllFieldsConstructor(Modifier.PUBLIC)
            .addSuperinterface(TypeDef.parameterized(ClassTypeDef.of(Supplier::class.java), TypeDef.parameterized(List::class.java, String::class.java).array())).addMethod(get).build()
        val result = TypeDef.parameterized(ClassTypeDef.of(Supplier::class.java), TypeDef.parameterized(List::class.java, Any::class.java).array())
        val caller = ClassDef.builder("test.ArraySupplierCaller")
            .addMethod(MethodDef.builder("reference").addParameter("target", target.asTypeDef()).returns(result)
                .build { _, p -> result.methodReference(p[0], get).returning() }).build()
        compileMatchingSnapshots(snapshots, target, caller).use { loader ->
            val targetClass = loader.loadClass(target.name)
            val cls = loader.loadClass(caller.name)
            val value = arrayOf(listOf("text"))
            val instance = targetClass.getConstructor(Any::class.java).newInstance(value as Any)
            val reference = cls.getMethod("reference", targetClass).invoke(cls.getConstructor().newInstance(), instance) as Supplier<*>
            assertSame(value, reference.get())
        }
    }

    @Test
    fun exhaustiveConditionalDropsTheUnreachableFallback() {
        val def = ClassDef.builder("test.ExhaustiveBranches")
            .addMethod(MethodDef.builder("call").addParameter("flag", TypeDef.Primitive.BOOLEAN).returns(String::class.java)
                .build { _, p -> StatementDef.multi(p[0].isTrue().doIfElse(
                    ExpressionDef.constant("yes").returning(), ExpressionDef.constant("no").returning()), ExpressionDef.constant("unreachable").returning()) }).build()
        compileMatchingSnapshots(snapshots, def).use { loader ->
            val cls = loader.loadClass(def.name)
            val instance = cls.getConstructor().newInstance()
            assertEquals("yes", cls.getMethod("call", Boolean::class.javaPrimitiveType).invoke(instance, true))
            assertEquals("no", cls.getMethod("call", Boolean::class.javaPrimitiveType).invoke(instance, false))
        }
    }

    class Calls {
        companion object {
            @JvmStatic
            fun <T : CharSequence> array1(value: Array<T>): Array<T> = value

            @JvmStatic
            fun <T : CharSequence> array2(value: Array<Array<T>>): Array<Array<T>> = value

            @JvmStatic
            fun <T : CharSequence> array3(value: Array<Array<Array<T>>>): Array<Array<Array<T>>> = value

            @JvmStatic
            fun <T : Number> number(value: T): T = value

            @JvmStatic
            fun <T : CharSequence> text(value: T): T = value

            @JvmStatic
            fun <T : List<String>> list(value: T): T = value
        }
    }
}
